lib/common/template.h

```diff
@@ -0,0 +1,288 @@
+      return N;
+    } else {
+      return SearchTypeListHelper<N + 1, PREDICATE, TUPLE>::value();
+    }
+  }
+};
+
+template<template<typename> class PREDICATE, typename... TYPES>
+constexpr int SearchTypeList{
+    SearchTypeListHelper<0, PREDICATE, std::tuple<TYPES...>>::value()};
+
+// TypeIndex<A, TYPES...> scans a list of types for simple type equality.
+// The zero-based index of A in the list is returned, or -1 if A is not present.
+template<typename A> struct MatchType {
+  template<typename B> struct Match {
+    static constexpr bool value() {
+      return std::is_same_v<std::decay_t<A>, std::decay_t<B>>;
+    }
+  };
+};
+
+template<typename A, typename... TYPES>
+constexpr int TypeIndex{SearchTypeList<MatchType<A>::template Match, TYPES...>};
+
+// IsTypeInList<A, TYPES...> is a simple presence predicate.
+template<typename A, typename... TYPES>
+constexpr bool IsTypeInList{TypeIndex<A, TYPES...> >= 0};
+
+// OverMembers extracts the list of types that constitute the alternatives
+// of a std::variant or elements of a std::tuple and passes that list as
+// parameter types to a given variadic template.
+template<template<typename...> class, typename> struct OverMembersHelper;
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::variant<Ts...>> {
+  using type = T<Ts...>;
+};
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::tuple<Ts...>> {
+  using type = T<Ts...>;
+};
+
+template<template<typename...> class T, typename TorV>
+using OverMembers = typename OverMembersHelper<T, std::decay_t<TorV>>::type;
+
+// SearchMembers<PREDICATE> scans the types that constitute the alternatives
+// of a std::variant instantiation or elements of a std::tuple.
+// The zero-based index of the first type T among the alternatives for which
+// PREDICATE<T>::value() is true is returned, or -1 when the predicate is false
+// for every type in the set.
+template<template<typename> class PREDICATE> struct SearchMembersHelper {
+  template<typename... Ts> struct Scanner {
+    static constexpr int value() { return SearchTypeList<PREDICATE, Ts...>; }
+  };
+};
+
+template<template<typename> class PREDICATE, typename TorV>
+constexpr int SearchMembers{
+    OverMembers<SearchMembersHelper<PREDICATE>::template Scanner,
+        TorV>::value()};
+
+template<typename A, typename TorV>
+constexpr bool HasMember{
+    SearchMembers<MatchType<A>::template Match, TorV> >= 0};
+
+// const A * -> std::optional<A>
+template<typename A> std::optional<A> GetIfNonNull(const A *p) {
+  if (p) {
+    return {*p};
+  }
+  return std::nullopt;
+}
+
+// const std::variant<..., A, ...> -> std::optional<A>
+// i.e., when a variant holds a value of a particular type, return a copy
+// of that value in a std::optional<>.  The type A must be a valid
+// alternative for the variant.
+template<typename A, typename VARIANT>
+std::optional<A> GetIf(const VARIANT &u) {
```

Comment:
What's the benefit of this over just using std::get_if? They seem almost interchangeable, but we have a lot of uses std::get_if already.

Reply:
I'll examine the uses of GetIf that survive and see whether they can be rewritten in terms of get_if.